lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,201 @@
+from flask import jsonify
+from logging import Logger
+from typing import Any, AnyStr, Dict
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources.query_apis import (
+    ElasticBase,
+    Schema,
+    Parameter,
+    ParamType,
+    PostprocessError,
+)
+
+
+class SearchAPI(ElasticBase):
+    """
+    Create a search request based on given query parameter.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("user", ParamType.USER, required=False),
+                Parameter("start", ParamType.DATE, required=True),
+                Parameter("end", ParamType.DATE, required=True),
+                Parameter("search_term", ParamType.STRING, required=False),
+                Parameter("fields", ParamType.LIST, required=False),
+            ),
+        )
+
+    def flatten_json(
+        self, es_json: Dict[AnyStr, Any], prefix="", out={}
+    ) -> Dict[AnyStr, Any]:
+        """
+        Utility function for flattening the json.
+        json_in =
+            {
+                    "@timestamp": "2021-07-12T14:08:46.563303",
+                    "run": {
+                        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
+                        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
+                    },
+                    "host_tools_info": [
+                        {
+                            "hostname": [{"test1":"dhcp31-171"}, {"test2":"dhcp31-195"}]
+                        }
+                    ]
+                }
+        json_out =
+            {
+                '@timestamp': '2021-07-12T14:08:46.563303',
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is dict:
+            for key in es_json:
+                while prefix[-2:-1] == ".":
+                    prefix = prefix[:-1]
+                self.flatten_json(es_json[key], prefix + key + ".", out)
+        elif type(es_json) is list:
+            i = 0
+            for element in es_json:
+                if type(element) is dict:
+                    self.flatten_json(element, prefix + ".", out)
+                else:
+                    self.flatten_json(element, prefix + str(i) + ".", out)
+                i += 1
```

Comment:
That is a valid point, but it would also mean that 
```
{"mylist": [{"key1": 'valueA'}, {"key2": 'valueB'}]}
```
converts to 
```
{"mylist.0.key1": 'valueA', "mylist.1.key2": 'valueB'}
```
Is this an appropriate output or we need to apply some more cleaning in the postprocessing?

Reply:
There are at least two definitions of "appropriate" here, and only one of them is subject to debate. (That means the first is somewhat objective, while the second/third/etc may be more fun.)

1. Produce exactly the same output for any Elasticsearch response as the current dashboard effect;
2. Figure out what we'd like it to do, and change the dashboard page code to deal with any difference from what the current effect produces. (Potentially many approaches here, including just redefining the API to *not* flatten.)

Even if there's a bug in the flattening of search results, it may well be that it never mattered because either there were no duplicate fields in `run` documents or the dashboard just never cared about any fields that were overwritten. In that case, it'd be nice to fix the bug without massive disruption in the page code if that's possible.

Reply:
I am not very sure about your first point because we don't want the output so tightly tied to Elasticsearch and we want to minimize the extra post processing that dashboard need to do. 
About the second point yes there are few fields in the `run` document that may have the same key name and thus may get overwritten for example `inet` and `inet6` of the `sorsreport` like following. However, I am not sure how the dashboard wants to filter using these fields. At the moment the dashboard returns 400 error when we specify the `sosreport.inet` or `sosreport.inet6` and that is correct because the coorect field name to select would be `sosreport.inet.<ifname/ipaddr>`  (which is something I should also test fr this PR). But are these fields you think dashboard never cared about and it can be overwritten? 
```
"sosreports": [
          {
              "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19/sysinfo/end/dhcp31-171/sosreport-dhcp31-171-pbench-dhcp31-171.perf.lab.eng.bos.redhat.com-2021-07-12-kpqchzn.tar.xz",
              "md5": "cda0d8e931b46e57fcc736165559fd74",
              "hostname-f": "dhcp31-171.perf.lab.eng.bos.redhat.com",
              "hostname-s": "dhcp31-171",
              "inet": [
                  {
                      "ifname": "lo",
                      "ipaddr": "127.0.0.1"
                  },
                  {
                      "ifname": "ens3",
                      "ipaddr": "10.16.31.171"
                  }
              ],
              "inet6": [
                  {
                      "ifname": "lo",
                      "ipaddr": "::1"
                  },
                  {
                      "ifname": "ens3",
                      "ipaddr": "2620:52:0:101c:5054:ff:fe5e:5a4a"
                  },
                  {
                      "ifname": "ens3",
                      "ipaddr": "fe80::5054:ff:fe5e:5a4a"
                  }
              ]
          }
      ]
```

Reply:
> Produce exactly the same output for any Elasticsearch response as the current dashboard effect

I suggest that this should not be a goal.  That is, I understand and appreciate the value of preserving the usefulness of existing code, but I would prefer to see us provide a solid and uniform foundation, and then rectify any problems that that poses for the stuff being built on top of it.  Moreover, bug for bug compatibility seems unwarranted in this situation, and, as Nikhil suggests, the ability to insulate the upper layers from the details of the database used in the lower layers seems like a Good Thing.

> there are few fields in the `run` document that may have the same key name and thus may get overwritten for example `inet` and `inet6` of the `sorsreport`

I'm not seeing how there could be conflicts between fields in `inet` and fields in `inet6` (since their respective prefixes will be different), but I do see in the example above a conflict _within_ `inet6`...but that leads me to a different question:  how does a single network interface ('ens3') come to have more than one `ipaddr`?  (That is, could this be demonstrating a bug in the `sosreport`?...shouldn't the third interface have a name which is different from the second's??)

As for how to flatten an array/list, I think we just need to state what our procedure is.  If we can guarantee/require that the array/list contains only objects and that none of the objects contain overlapping keys, then we can simply merge the list to produce a single object; however, if we cannot guarantee or require that, then we either have to insert the element index as a component in the prefix, to ensure that no values get overridden, or we need to state that duplicate keys will result in values being lost or overridden (or that they will result in errors).

I can imagine how having the loop index inserted into the name would be highly undesirable, as the value might be arbitrary or change in ways which are unpredictable.  But, I can also imagine that having values overridden would also be highly annoying, but I think that that is unavoidable without some sort of arbitrary disambiguation, which is what the index-insertion provides in a reasonable way.

Reply:
My suggestion for the immediate future would be to write the code to merge the list into a single object and have it raise an error if the list contains anything which is not an object and if merging any of the objects results in a name-collision.

That is, for the moment, lets assume that the purpose here is to merge a list of objects which have no overlap, and therefore we should complain bitterly if asked to do anything else.  If and when we discover that we have a requirement to do something else, then we can grapple with how to handle it.  (The approach in the code as it stands seems like the worst of all worlds -- the decision of whether to insert the element index into the prefix should be consistent for all elements in a given list and should be made based on the entire list contents -- it shouldn't be different for different elements.)

Reply:
Strategically, one option is to stop flattening the results entirely and return the original hierarchy. Flattening was an artifact of the ES1 query response. To simplify the dashboard changes when we went to ES7 last fall, we flattened the ES7 `_source` to resembled the ES1 response and isolate the dashboard changes to the service and model layer.

As for the IP duplication ... well, it's interesting the way the SOS report gathers all of the `inet` data across interfaces into one bundle and the `inet6` data across interfaces into another. So the `lo` interface has an IPv6 and an IPv5, while the `ens3` interface has an IPv4 address and two IPv6 addresses; all of which is entirely reasonable. The problem here is solely in the somewhat odd concept of "flattening" the array elements. Again, flattening as `sosreports.inet6.0.ifname` and `sosreports.inet6.1.ifname` would solve the overlay problem, but would presumably be more difficult to decode in the page logic.

Retaining the original structure would make a lot more sense, but I have no idea how much work would need to be done in the results page display to handle it.

Reply:
> Retaining the original structure would make a lot more sense, but I have no idea how much work would need to be done in the results page display to handle it.

I mean for the result page, someone has to flatten the data, either the server or the dashboard. If the dashboard wants the flattened view of the json data for all it's purpose then it make sense for us to return it otherwise if the dashboard does want the json view of the data for its other purpose such as run comparison or something then it make sense for us to return the json view of the data

Reply:
The page component needs to *display* the data returned. Yes, the search results view uses the flattened field paths as column headers in the table, but Javascript is perfectly capable of handling that itself. The important question becomes how we want the server to represent the data. I don't think `sosreports.inet6.1.ifname` is particularly useful for the dashboard, for example; that's not going to be a column heading... instead of having a column for each item in each inet array, you'd lump them all under `sosreports.inet6` (which currently blows up on the production server, amusingly). In this case, I think the structured JSON data is much more useful than the flattened form.

Reply:
So you are suggesting we return the ES result as is? and let the dashboard transform it

Reply:
> the `ens3` interface has an IPv4 address and two IPv6 addresses

I'm still hung up on the notion that a single interface could have more than one address...I don't think that that is right...and, if that were to get fixed, then we might be able to finesse the problem.  We can flatten things OK so long as there aren't any arrays or so long as the elements in the array result in unique keys...thus, if the only non-unique key turns out to be a bug, then we nail two wins....

Can anybody state whether the sosreport is right or wrong?

Reply:
Is this some sort of aggregation returned by ES, such that the same interface (`ens3`) on _different_ systems has a different address?  _That_ would make sense...but, then, why would the key and value both be ending up in the flattened name?

Reply:
@webbnh A NIC can only have *one* IPv4 address, that's true; but the IPv6 spec allows any NIC to have any number of IPv6 addresses. In fact it's virtually impossible not to have at least 2 (a routable address and a link-local address, which is the `fe80:` prefix). In most contexts, a link-local address isn't very useful (because it's only reachable on the local subnet), but in others it's critical. However, we could also have multiple routable addresses. IPv6 is fascinating; too bad it remains so uncommon after all this time.

Reply:
@npalaska Returning the raw Elasticsearch is certainly one option. The dashboard model already has code to transform it, and we'd still be achieving the basic goal of isolating the client interface from our back-end database. Another would be to devise a new format we like better.

I'm only a little concerned about our `sosreport` duplication here, because that appears to be an existing 0.69 problem so we're probably at least not making things worse ... but the resulting structure is *not* useful and won't be handled right by the dashboard. On the other hand, that means we need to change the dashboard code either way.

Once we have `queryIndexMapping` fixed, I can experiment with this a bit more in the dashboard code and get more information to fuel the discussion.

Reply:
Thanks @webbnh and @dbutenhof  for some interesting points on ipv4 and ipv6 :) But yeah coming back to this issue we need to finalize on the output formatting. I do think the flattened structure is not entirely _not useful_ it would be like "pick and place" for dashboard something like grab a key and a value for that key from the list and display it on the search pane without having to do any sort of postprocessing. However, I too am a little concerned about overriding as Webb mentioned and not sure how important those duplicate fields are from the dashboard perspective. But I also tend to agree with Dave that entering _0s_ or _1s_ or _Ns_ in  our dot separated nested key values could be little confusing. And Dave If you want do some experiments once `queryIndexMapping` is merged that would be great.

Reply:
My concern with the enumerated entries is not that they would be confusing, but rather that they might produce inconsistent results.  That is, unless the array entries are somehow strongly ordered, a given value would sometimes appear at entry _X_ and sometimes at entry _Y_ which at best might produce a poor UX and at worst might defeat down-line aggregations or other organization.

However, that said, I think that "losing" values is completely unacceptable.  So, if two values correspond to the same prefix/key, then either the keys need to be disambiguated some how, or it has to result in an error...and, in the general case, _neither_ of those solutions is likely to be adequate.

So, if we can convince ourselves that (or if we can massage the input such that) there should never be key collisions (that is, if we can take ourselves out of "the general case"), then I'm fine with continuing to flatten the keyspace and with doing so without enumerating the array entries (and with having collisions result in errors).  However, if we know that we're going to be stuck with things like the `sosreport` duplication, then I think we need to seek an alternate solution which is not flat.

Dave says that the Dashboard code is going to have to be modified regardless of the choice made here, so that frees our hand considerably.  Simply returning the ES result more or less "as is" is very attractive; however, I think it would be prudent to assume that we might not always be using ES to store this information, and therefore a reasonable, database-independent format would be better.

A nice compromise approach might be to return a structure which is "mostly flat".  That is, if we were to permit lists at the leaf-level, I think that could allow us to solve the problem.  That is, we can continue to use an approach similar to the current one, but, when we store a value for a given flattened key, if the key already exists, then we convert the existing value to a list and add the new value to it.  This should allow the client to enjoy the features of a flat name-space without losing colliding entries (and, it's just up to the client to handle things appropriately).

(Is there a (re)solution in there somewhere?)